src/mono/netcore/System.Private.CoreLib/src/System/Reflection/Emit/AQNBuilder.Mono.cs

```diff
@@ -0,0 +1,483 @@
+using System.Text;
+
+namespace System.Reflection.Emit
+{
+	// TypeNameBuilder is NOT thread safe NOR reliable
+	internal class TypeNameBuilder
+	{
+		internal enum Format
+		{
+			ToString,
+			FullName,
+			AssemblyQualifiedName,
+		}
+
+		private static class TypeString
+		{
+			public static bool  ContainsReservedChar (string name)
+			{
+				foreach (char c in name.AsSpan ())
+				{
+					if (c == '\0')
+						break;
+					if (TypeString.IsTypeNameReservedChar (c))
+						return true;
+				}
+				return false;
+			}
+
+			public static bool IsTypeNameReservedChar(char ch)
+			{
+				switch (ch)
+				{
+				case ',':
+				case '[':
+				case ']':
+				case '&':
+				case '*':
+				case '+':
+				case '\\':
+					return true;
+
+				default:
+					return false;
+				}
+			}
+		}
+
+		private class TypeNameBuilderImpl : IDisposable
+		{
+			private enum ParseState
+			{
+				START         = 0x0001,
+				NAME          = 0x0004,
+				GENARGS       = 0x0008,
+				PTRARR        = 0x0010,
+				BYREF         = 0x0020,
+				ASSEMSPEC     = 0x0080,
+				ERROR         = 0x0100,
+			 }
+
+			ParseState m_parseState;
+			StringBuilder m_pStr = new StringBuilder ();
+			int m_instNesting;
+			bool m_bFirstInstArg;
+			bool m_bNestedName;
+			bool m_bHasAssemblySpec;
+			bool m_bUseAngleBracketsForGenerics;
+			Stack m_stack = new Stack ();
+
+			public void Dispose () { m_pStr = null; }
+			public void OpenGenericArguments()
+			{
+				if (!CheckParseState(ParseState.NAME))
+					Fail();
+
+				m_parseState = ParseState.START;
+				m_instNesting ++;
+				m_bFirstInstArg = true;
+
+				if (m_bUseAngleBracketsForGenerics)
+					Append('<');
+				else
+					Append('[');
+
+			}
+			public void CloseGenericArguments()
+			{
+				if (m_instNesting == 0)
+					Fail();
+				if (!CheckParseState(ParseState.START))
+					Fail();
+
+				m_parseState = ParseState.GENARGS;
+
+				m_instNesting --;
+
+				if (m_bFirstInstArg)
+				{
+					m_pStr.Remove(m_pStr.Length - 1, 1);
+				}
+				else
+				{
+					if (m_bUseAngleBracketsForGenerics)
+						Append('>');
+					else
+						Append(']');
+				}
+			}
+			public void OpenGenericArgument()
+			{
+				if(!CheckParseState (ParseState.START))
+					Fail ();
+
+				if (m_instNesting == 0)
+					Fail ();
+
+				m_parseState = ParseState.START;
+				m_bNestedName = false;
+
+				if (!m_bFirstInstArg)
+					Append(',');
+
+				m_bFirstInstArg = false;
+
+				if (m_bUseAngleBracketsForGenerics)
+					Append('<');
+				else
+					Append('[');
+				PushOpenGenericArgument();
+			}
+
+			public void CloseGenericArgument()
+			{
+				if (!CheckParseState(ParseState.NAME | ParseState.GENARGS | ParseState.PTRARR | ParseState.BYREF | ParseState.ASSEMSPEC))
+					Fail();
+
+				if (m_instNesting == 0)
+					Fail();
+
+				m_parseState = ParseState.START;
+
+				if (m_bHasAssemblySpec)
+				{
+					if (m_bUseAngleBracketsForGenerics)
+						Append('>');
+					else
+						Append(']');
+				}
+
+				PopOpenGenericArgument();
+			}
```

Comment:
Nit: some methods have blank lines in between them, some don't.  We generally do in the repo.  It'd be good to be consistent, at least within the file :)